src/openrct2/network/Network.cpp

```diff
@@ -1942,6 +1974,80 @@ void Network::ProcessPending()
     ProcessPlayerList();
 }
 
+static bool ProcessPlayerAuthenticatePluginHooks(
+    const NetworkConnection& connection, const std::string_view& name, const std::string_view& publicKeyHash)
+{
+#    ifdef ENABLE_SCRIPTING
+    using namespace OpenRCT2::Scripting;
+
+    auto& hookEngine = GetContext()->GetScriptEngine().GetHookEngine();
+    if (hookEngine.HasSubscriptions(OpenRCT2::Scripting::HOOK_TYPE::NETWORK_AUTHENTICATE))
+    {
+        auto ctx = GetContext()->GetScriptEngine().GetContext();
+
+        // Create event args object
+        DukObject eObj(ctx);
+        eObj.Set("name", name);
+        eObj.Set("publicKeyHash", publicKeyHash);
+        eObj.Set("ipAddress", connection.Socket->GetIpAddress());
+        eObj.Set("cancel", false);
+        auto e = eObj.Take();
+
+        // Call the subscriptions
+        hookEngine.Call(OpenRCT2::Scripting::HOOK_TYPE::NETWORK_AUTHENTICATE, e, false);
+
+        // Check if any hook has cancelled the join
```

Comment:
Does this mean if the user started authentication and left before actually doing so?

Reply:
No any of the hook event handlers can deny the user from entry. The same object is used for each event handler and they can fight over the cancel boolean amongst themselves. We then take the final result of cancel and if it is true, we deny the user from entry into the server. This is before they have "official" joined and been given a player id.

Reply:
I see, is the concurrency expected or something we'll have to deal with later ?

Reply:
It isn't concurrent, all very synchronous. Yes it is by design.